### ISSUE TITLE
Crash during runloop's autorelease pool drain on macOS 10.15 Catalina

### ISSUE THREAD

**Issue by iby** [ReactiveCocoa/ReactiveCocoa#3690]
Experiencing a naughty crash with buttons and all inheriting classes inside NSStackView with enabled `Detaches Hidden Views` when the window gets closed and deallocated in the latest Catalina build:

```
2019-10-06 15:04:27.548688+0100 App[4474:58829] [general] Caught exception during runloop's autorelease pool drain of client objects NSRangeException: Cannot remove an observer <NSStackView 0x10461aca0> for the key path "hidden" from <NSButton 0x10461c7f0> because it is not registered as an observer. userInfo: (null)
2019-10-06 15:04:27.550252+0100 App[4474:58829] [General] An uncaught exception was raised
2019-10-06 15:04:27.550313+0100 App[4474:58829] [General] Cannot remove an observer <NSStackView 0x10461aca0> for the key path "hidden" from <NSButton 0x10461c7f0> because it is not registered as an observer.
	0   CoreFoundation                      0x00007fff36166d63 __exceptionPreprocess + 250
	1   libobjc.A.dylib                     0x00007fff6bf56bd4 objc_exception_throw + 48
	2   Foundation                          0x00007fff387a7395 -[NSObject(NSKeyValueObserverRegistration) _removeObserver:forProperty:] + 578
	3   Foundation                          0x00007fff387a7106 -[NSObject(NSKeyValueObserverRegistration) removeObserver:forKeyPath:] + 74
	4   Foundation                          0x00007fff387bf963 -[NSObject(NSKeyValueObserverRegistration) removeObserver:forKeyPath:context:] + 190
	5   AppKit                              0x00007fff334e0cc9 -[NSStackView dealloc] + 233
	6   CoreFoundation                      0x00007fff360c5176 -[__NSArrayI dealloc] + 73
	7   libobjc.A.dylib                     0x00007fff6bf6653a _ZN19AutoreleasePoolPage12releaseUntilEPP11objc_object + 134
	8   libobjc.A.dylib                     0x00007fff6bf4cc70 objc_autoreleasePoolPop + 175
	9   CoreFoundation                      0x00007fff3608767e _CFAutoreleasePoolPop + 22
	10  CoreFoundation                      0x00007fff360b59fc __CFRunLoopRun + 2404
	11  CoreFoundation                      0x00007fff360b4e13 CFRunLoopRunSpecific + 499
	12  HIToolbox                           0x00007fff34c41b2d RunCurrentEventLoopInMode + 292
	13  HIToolbox                           0x00007fff34c4186d ReceiveNextEventCommon + 600
	14  HIToolbox                           0x00007fff34c415f7 _BlockUntilNextEventMatchingListInModeWithFilter + 64
	15  AppKit                              0x00007fff332eaac4 _DPSNextEvent + 990
	16  AppKit                              0x00007fff332e9834 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1352
	17  AppKit                              0x00007fff332e3fd4 -[NSApplication run] + 658
	18  AppKit                              0x00007fff332d5e7d NSApplicationMain + 777
```

As far as I can tell this has been happening since Catalina Beta was released and still occurs on the latest GM seed, which is worrying… Builds produced with the latest stable Xcode and Xcode 11.2 Beta suffer from the same problem. Everything pre-Catalina works 100% fine.

All of the above narrows down to ReactiveCocoa bindings. Removing the following `observeValues` line fixes the issue:

```swift
override func viewDidLoad() {
     self.launchOnStartupCheckbox.reactive.boolValues.observeValues({ Swift.print($0) })
}
```

It seems like there's something wrong with ObjC swizzling that ReactiveCocoa does behind the scenes? It also looks there're some observer-related issues, is this something related or already known? Otherwise happy to try isolating the issue into a sample project, but there's really not much details besides that: NSStackView with enabled `Detaches Hidden Views` + NSButton + reactive observer on button's IBOutlet = 💥 on window deallocation!

**Comment by ianbytchek** [comment]
With Catalina out in the wild could you comment on the above?

**Comment by ohbargain** [comment]
I hit this problem. @ianbytchek were you able to work around this problem?

**Comment by ianbytchek** [comment]
@ohbargain I disabled `Detaches Hidden Views` on all stack views where it wasn't needed, apparently that's a good thing anyway, i.e., less observers = less resources. In my case this wasn't needed in over 95% cases. For the rest I reverted to `@IBActions` in favour of reactive observers for the simplicity sake. But you can also wrap these controls into another view and hide it instead, so the NSStackView would observe the wrapper and the ReactiveCocoa would swizzle the control itself without causing a conflict between the two.

**Comment by andersio** [comment]
This is quite an interesting issue, because RAC would not have swizzled things that affect KVO observer registration. 🤔

**Comment by ianbytchek** [comment]
Doesn't seem to be happening any longer on latest Catalina 10.15.4. I encountered a few other notorious AppKit crashes on 10.15, so this probably was one of these internal oversights. Feel free to close.

**Comment by ianbytchek** [comment]
Sorry, false negative. I must have tested the piece without any bindings.

**Comment by ianbytchek** [comment]
@andersio FYI, you seem to have already worked on the related #3433 issue a few years back and fixed it in #3435. Does that ring the bell? I tried poking the codebase but I feel completely out of context with the swizzling part… :(

**Comment by ianbytchek** [comment]
Another interesting fact. If I understand correctly ReactiveCocoa uses different swizzling for different things. So, depending on that and on the observation order it shows different results. There are cases when turning off hidden view detaching doesn't even work.

```swift
extension Reactive where Base: NSView {
    /// Returns a signal for the given key path using native AppKit block observation.
    internal func _signal<U>(`for` keyPath: KeyPath<Base, U>) -> Signal<Bool, Never> {
        Signal({ rxObserver, lifetime in
            let nsObserver = self.base.observe(keyPath, changeHandler: { r, _ in rxObserver.send(value: r.isHidden) })
            lifetime.observeEnded({ let _ = nsObserver }) // Retain observer for the lifetime of signal.
            lifetime += self.lifetime.observeEnded(rxObserver.sendCompleted) // Complete signal when the object released.
        })
    }
}

autoreleasepool {
    let w = NSWindow()
    let s = NSStackView()
    let b = NSButton()
    s.addArrangedSubview(b)
    w.contentView?.addSubview(s)

    // Single observer: 👍
    // s.detachesHiddenViews = true
    // b.reactive.signal(for: \.isHidden).observeValues({ print("isHidden:", $0) })
    
    // Single observer: 💣
    // s.detachesHiddenViews = true
    // b.reactive.boolValues.observeValues({ print("boolValue:", $0) })
    
    // Single observer: 👍
    // s.detachesHiddenViews = false
    // b.reactive.boolValues.observeValues({ print("boolValue:", $0) })
    
    // Multiple observers: 👍
    // s.detachesHiddenViews = false
    // b.reactive.boolValues.observeValues({ print("boolValue:", $0) })
    // b.reactive.signal(for: \.isHidden).observeValues({ print("isHidden:", $0) })
    
    // 🔥🔥🔥 Same as above, different order: 💣
    s.detachesHiddenViews = false
    b.reactive.signal(for: \.isHidden).observeValues({ print("isHidden:", $0) })
    b.reactive.boolValues.observeValues({ print("boolValue:", $0) })
    
    // Using custom observer: 👍
    // s.detachesHiddenViews = false
    // b.reactive._signal(for: \.isHidden).observeValues({ print("isHidden:", $0) })
    // b.reactive.boolValues.observeValues({ print("boolValue:", $0) })

    b.performClick(nil)
    b.performClick(nil)
    b.isHidden = true
    b.isHidden = false
}
```

**Comment by ianbytchek** [comment]
😅 Using a custom observer leads to a whole different crash.

```swift
#0	0x00007fff689ff8fe in object_isClass ()
#1	0x00007fff322b8348 in KVO_IS_RETAINING_ALL_OBSERVERS_OF_THIS_OBJECT_IF_IT_CRASHES_AN_OBSERVER_WAS_OVERRELEASED_OR_SMASHED ()
#2	0x00007fff32456301 in NSKeyValueWillChangeWithPerThreadPendingNotifications ()
#3	0x00007fff2cf02f35 in -[NSView _viewDidChangeAppearance:] ()
#4	0x00007fff2cf0312f in -[NSView _viewDidChangeAppearance:] ()
#5	0x00007fff2ceeaf8c in -[NSView initWithCoder:] ()
#6	0x00007fff2cf3f9f7 in -[NSStackView initWithCoder:] ()
#7	0x00007fff322a2ae0 in _decodeObjectBinary ()
#8	0x00007fff322a1f40 in _decodeObject ()
#9	0x00007fff322a1dd4 in -[NSKeyedUnarchiver decodeObjectForKey:] ()
#10	0x00007fff2ceeb32b in -[NSResponder initWithCoder:] ()
#11	0x00007fff2cee9b47 in -[NSView initWithCoder:] ()
#12	0x00007fff2cf410a6 in -[NSControl initWithCoder:] ()
#13	0x00007fff2cf40fe6 in -[NSButton initWithCoder:] ()
#14	0x00000001002e20fe in BooleanSetting.init(coder:)
…
```

This is a random crash and doesn't necessarily happen in the same place, occurs when loading a view from a storyboard and creating fresh binding for the 3rd-4th time. I'm guessing it has to do with retaining `nsObserver` and releasing it at around the same time when the base object gets deallocated.

**Comment by andersio** [comment]
Could you share at which point you are creating the bindings? It feels like there is a possibility of `NSSecureCoding` checks having kicked in, when the instance was swizzled before the `NSSecureCoding` process is completed.

On paper though, crashing at this symbol likely means we have undeteched observers that weren't covered by 10.13+ KVO auto-removal. 🤔

**Comment by ianbytchek** [comment]
Of course! The original crash with extra cases is fully covered in the [comment above](https://github.com/ReactiveCocoa/ReactiveCocoa/issues/3690#issuecomment-626319538), it can be copy-pasted in the ReactiveCocoa AppKit playground. I didn't find a way to show full debug info inside the playground, but you can observe the same issue in [NSbuttonSpec](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/aef73316518c43063a091b625d263ebabd177fb0/ReactiveCocoaTests/AppKit/NSButtonSpec.swift#L81-L117), which currently fails as mentioned by @simba909. These bindings are made inside `viewDidLoad()` in `NSViewController` and inside `awakeFromNib` in `NSView`.

As for the `KVO_IS_RETAINING_ALL_…_OBSERVER_WAS_OVERRELEASED_OR_SMASHED` crash, it's not very clear how much of it is caused by ReactiveCocoa. Tt doesn't occur every single time and the stack trace isn't always the same, sometimes containing ReactiveCocoa/Swift frames, sometimes not. However:

```swift
Signal({ rxObserver, lifetime in
    let nsObserver = self.base.observe(keyPath, changeHandler: { r, _ in rxObserver.send(value: r.isHidden) })
    // 👉 If inside here I retain the self (+base), the crash stops happening, but this leads to
    // 👉 a retain cycle unless the disposable is explicitly destroyed.
    lifetime.observeEnded({ let _ = (nsObserver, self) }) // Retain observer for the lifetime of signal.
    lifetime += self.lifetime.observeEnded(rxObserver.sendCompleted) // Complete signal when the object released.
})
```

If I understand correctly, the `…OBSERVER_WAS_OVERRELEASED_OR_SMASHED` is caused by unbalanced observers, i.e., removing an observer that was never added. Though, [NSKeyValueObservation](https://github.com/apple/swift/blob/7a238c633ae195902b4a164f69ca1cabb3398c72/stdlib/public/Darwin/Foundation/NSObject.swift#L163-L238) is supposed to handle everything automatically behind the scenes it also uses swizzling, which might be conflicting with what ReactiveCocoa does. I would also make these bindings inside `viewDidLoad()` in `NSViewController` and inside `awakeFromNib` in `NSView`. If I can be useful with hunting this down, please do tell!

**Comment by andersio** [comment]
@ianbytchek This looks like isa-swizzling RAC did is intefering with some assumptions made by NSStackView internals in Catalina. Have to look deeper into how this can be resolved.

**Comment by andersio** [comment]
It seems the issue is around that we isa-swizzle the class **after** the internal NSKeyValueProperty instance was created, which had cached the isa pointer to the original `NSButton` class. But then at removal time, it was perhaps attempting to match the observer using the new isa pointer to our swizzled class.

:|

They have likely tightened how KVO detects isa swizzling.

**Comment by ianbytchek** [comment]
Interesting! Does it mean that removing the view from the `NSStackView`, adding reactive observers, and adding it to view might solve the issue? Or, alternatively, swizzling the view at the `init` stage to achieve the same?

**Comment by andersio** [comment]
Make sure RAC swizzling has happened before adding the `NSButton` to a `NSStackView` would likely workaround the issue reliably.

**Comment by andersio** [comment]
Make sure all RAC bindings that would swizzle have setup, before adding the NSViews to a NSStackView would workaround the issue reliably.

**Comment by andersio** [comment]
Alright, I managed to figure out one of the cause. RAC deliberately avoided this situation by not isa-swizzling if KVO has isa-swizzled already (so we don't subclass the KVO `NSKVONotifying_` runtime subclass).

That's said Catalina appears to have broken it because `object_getClass` is now conceiving `NSKVONotifying_`, when it had been reflecting the `isa` honestly before 10.15 Catalina AFAIK. 🤷‍♂️

**Comment by andersio** [comment]
Since it affects only macOS Catalina for now, the fix I have in mind would be the RAC internal `ActionProxy` swizzling the `NSControl` directly, instead of isa-swizzling. That hopefully would solve the issue around using with `NSStackView`, and should not change the behavior of things depending on `ActionProxy`.

Having that said, this might reemerge with method wiretapping with `trigger(for:)` and `signal(for:)`, unfortunately. Changing them not to do isa-swizzling would require a RAC major release, because we can no longer guarantee that the callout runs after the call (including the recursive `super` calls), hence potentially leading to unintended behavioral change. :(

**Comment by ianbytchek** [comment]
Thanks for digging into this @andersio. I use `trigger(for:)` and `signal(for:)` a lot and concerned whether this solution is swapping one issue for another?  

> Having that said, this might reemerge with method wiretapping with trigger(for:) and signal(for:), unfortunately. Changing them not to do isa-swizzling would require a RAC major release with a precaution notice, because we can no longer guarantee that the method invocation callout happens after the actuall call (including the recursive super calls), hence potentially leading to unintended behavioral change. :(

If that happens, would it be because of some other code potentially swizzled the "already directly swizzled `NSControl`", i.e., it's my responsibility that nothing messes with it, or can there be other situations that lead to problems?

**Comment by andersio** [comment]
@ianbytchek The issue stands on its own in an environment having only RAC. So I am fairly confident it is not a user-end error.

**Comment by ianbytchek** [comment]
Thanks for the update @andersio! Not sure if I follow, though. Are you saying this is more of an implementation problem/challenge, and not something for the end user to be concerned about?

> We probably also need field reports/tests on how RAC is doing with iOS 14 and macOS 11 as min deployment, since both contain some new Objective-C runtime changes.

True. If I come across anything else I'll certainly let know.

**Comment by RuiAAPeres** [comment]
Hello. 👋 Thanks for opening this issue. Due to inactivity, we will soft close the issue. If you feel that it should remain open, please let us know. 😄

**Comment by iby** [comment]
Unfortunately, I'm seeing the exact same exception with NSWindow in macOS 15 Sequoia now… I wonder if there's anything we can do about this. Clearly there's more and more changing in Cocoa, and things stop working at random.

```
Cannot remove an observer <NSKVOForwarder 0x600002c4e170> for the key path "_windowLayerContext" from <App.DialogWindow 0x134620e30> because it is not registered as an observer.
(
    0   CoreFoundation                      0x0000000193286920 __exceptionPreprocess + 176
    1   libobjc.A.dylib                     0x0000000192d6eb1c objc_exception_throw + 76
    2   Foundation                          0x00000001943c6544 -[NSObject(NSKeyValueObserverRegistration) _removeObserver:forProperty:] + 628
    3   Foundation                          0x00000001943c627c -[NSObject(NSKeyValueObserverRegistration) removeObserver:forKeyPath:] + 136
    4   Foundation                          0x00000001943c5f50 -[NSObject(NSKeyValueObserverRegistration) removeObserver:forKeyPath:context:] + 208
    5   AppKit                              0x0000000197914cfc -[NSWindow removeObserver:forKeyPath:context:] + 104
    6   AppKit                              0x00000001975b809c -[NSObject _removeForwardedObserver:forKeyPath:context:] + 104
    7   AppKit                              0x0000000196f2da6c -[NSWindow dealloc] + 248
    8   App.debug.dylib                     0x0000000106583750 $s13ReactiveSwift8LifetimeC0A5CocoaE2ofyACSo8NSObjectCFZACyXEfU_yyXEfU_ySVcfU_ + 796
    9   App.debug.dylib                     0x0000000106583798 $sSVIegy_SVIeyBy_TR + 56
    10  libobjc.A.dylib                     0x0000000192d63c44 _ZN19AutoreleasePoolPage12releaseUntilEPP11objc_object + 204
    11  libobjc.A.dylib                     0x0000000192d60158 objc_autoreleasePoolPop + 260
    12  CoreFoundation                      0x00000001931d28b8 _CFAutoreleasePoolPop + 32
    13  Foundation                          0x00000001943a8058 -[NSAutoreleasePool drain] + 140
    14  AppKit                              0x0000000196d2e82c -[NSApplication run] + 528
    15  AppKit                              0x0000000196d04fa4 NSApplicationMain + 888
    16  App.debug.dylib                     0x0000000106306774 $s5App10entrypoints5Int32VyF + 16
    17  App.debug.dylib                     0x00000001062a1e10 $s5App10EntrypointC5starts5Int32VyFZ + 1392
    18  App.debug.dylib                     0x00000001062a22f0 $s5App10EntrypointC5starts5Int32VyFZTo + 40
    19  App.debug.dylib                     0x0000000106201784 __debug_main_executable_dylib_entry_point + 28
    20  dyld                                0x0000000192dab274 start + 2840
)
```

The following:

```swift
let w = NSWindow()
 _ = w.reactive.trigger(for: #selector(NSWindow.performClose(_:)))
```

Crashes here: https://github.com/ReactiveCocoa/ReactiveCocoa/blob/f2d9bd56ae9f345821d9cd53fe3479db77e29094/ReactiveCocoa/AnyObject%2BLifetime.swift#L95

Like in my [old comment](https://github.com/ReactiveCocoa/ReactiveCocoa/issues/3690#issuecomment-626335151), simply adding a key-path observation first fixes this…

```swift
let w = NSWindow()
_ = w.reactive.signal(forKeyPath: "title")
_ = w.reactive.trigger(for: #selector(NSWindow.performClose(_:)))
```

@andersio @RuiAAPeres I understand that ReactiveCocoa is not in the "active development" phase any more. I just wonder if we can do these simple hack-arounds per macOS version for known classes to avoid these exceptions? Or if we know it's the swizzle-order, maybe we can just force-swizzle in the right order on all objects.

**Comment by iby** [comment]
After banging head on the wall trying to figure out a universal cure for this and making the code more maintainable and predictable (I don't want a million fixes for this all over different files and modules), I came up with (what seems) a pretty neat workaround – It involves some homeopathy and fixing swizzling issues with more swizzling… 👹🥲

As @andersio mentioned earlier, the problems start transpiring when doing [ISA-swizzling](https://stackoverflow.com/a/38878119/458356) **after** an object gets key-value-observed. The solution is very simple in theory, but not in practice: ISA-swizzle instances **before** the observation takes place.

A nice workaround is to simply override observation and trigger [ReactiveCocoa ISA-swizzling](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/12.0.0/ReactiveCocoa/NSObject%2BIntercepting.swift#L63), but this works only in a handful of cases :

```swift
override func addObserver(_ observer: NSObject, forKeyPath keyPath: String, options: NSKeyValueObservingOptions = [], context: UnsafeMutableRawPointer?) {
    _ = self.reactive.signal(for: #selector(Base.description))
    super.addObserver(observer, forKeyPath: keyPath, options: options, context: context)
}
```

A generalized approach requires class method swizzling and handling that interception there. I was very reluctant to do this at a global level as, theoretically, it can break something else.

But because I know exactly what causes issues in my code, I made it very selective with pre-swizzling-KVO for certain classes in certain situations:

```swift

// ☝️ You need this only if you're using `signal(for: Selector)` or `trigger(for Selector)`.
// ☝️ Here, I know that in my code only some controls inside stack views are causing issues,
// ☝️ so I only pre-swizzle those if they're inside a stack view – your situation may be different.

if #available(macOS 10.15, *) { NSControl.reactive.preSwizzleKVO({ $0.superview is NSStackView }) }
if #available(macOS 15, *) { NSWindow.reactive.preSwizzleKVO() }

extension NSObject {
    /// Intercepts the specified selector invocation by adding / replacing its method implementation. The interception
    /// handling is defined by the specified the block that takes the current method implementation and its selector,
    /// and returns a `@convention(block)` that becomes the new implementation.
    ///
    /// IMPORTANT: The `Implementation` and `Method` signatures must match – there's no safe way to check them and,
    /// unfortunately, generic params can't be used with Objective-C functions.
    ///
    ///   typealias Implementation = @convention(c) (Unmanaged<NSObject>, Selector, Arg1, Arg2, Arg3, ...) -> T
    ///   typealias Method = @convention(block) (Unmanaged<NSObject>, Arg1, Arg2, Arg3, ...) -> T
    ///   self.intercept(selector: #selector(...), with: { (implementation: Implementation) -> Method in
    ///     { reference, arg1, arg2, arg3, ... in
    ///       print("Intercepted before:", reference.takeUnretainedValue())
    ///       implementation(reference, selector, observer, keyPath, options, context)
    ///       // Custom after…
    ///     }
    ///   })
    static fileprivate func intercept<Implementation, Method>(selector: Selector, with implementation: (Implementation, Selector) -> Method) {
        guard let method = class_getInstanceMethod(self, selector) else { fatalError() }
        guard let encoding = method_getTypeEncoding(method) else { fatalError() }
        let oldImplementation = method_getImplementation(method)
        let newImplementation = imp_implementationWithBlock(implementation(unsafeBitCast(oldImplementation, to: Implementation.self), selector))
        if !class_addMethod(self, selector, newImplementation, encoding) {
            method_setImplementation(method, newImplementation)
        }
    }
}

extension Reactive where Base: NSObject {
    /// Pre-ISA-swizzle an instance when it gets key-value-observed with an optional filter.
    public static func preSwizzleKVO(_ filter: ((_ observee: Base, _ observer: NSObject, _ keyPath: String) -> Bool)? = nil) {
        typealias Implementation = @convention(c) (Unmanaged<NSObject>, Selector, NSObject, String, NSKeyValueObservingOptions, UnsafeMutableRawPointer?) -> Void
        typealias Method = @convention(block) (Unmanaged<NSObject>, NSObject, String, NSKeyValueObservingOptions, UnsafeMutableRawPointer?) -> Void
        Base.intercept(selector: #selector(Base.addObserver(_:forKeyPath:options:context:)), with: { (implementation: Implementation, selector:Selector) -> Method in
            { reference, observer, keyPath, options, context in
                let needsPreSwizzle = filter?(reference.takeUnretainedValue() as! Base, observer, keyPath) ?? true
                if needsPreSwizzle { _ = reference.takeUnretainedValue().reactive.signal(for: #selector(Base.description)) }
                implementation(reference, selector, observer, keyPath, options, context)
            } as Method
        })
    }

    public static func preSwizzleKVO(_ filter: @escaping () -> Bool) { self.preSwizzleKVO({ _, _, _ in filter() }) }
    public static func preSwizzleKVO(_ filter: @escaping (_ observee: Base) -> Bool) { self.preSwizzleKVO({ observee, _, _ in filter(observee) }) }
    public static func preSwizzleKVO(_ filter: @escaping (_ observee: Base, _ observer: NSObject) -> Bool) { self.preSwizzleKVO({ observee, observer, _ in filter(observee, observer) }) }
}
```

This allows to solve the problem globally for the app with a few lines at launch. Here's my Gist with relevant notes and playground: [iby / ISA-swizzling and breaking KVO](https://gist.github.com/iby/7db7dad3fc428a54b61838e022ceafe2)
